src/modules/performance_collector/db.py

```diff
@@ -0,0 +1,168 @@
+import sqlite3
+from contextlib import contextmanager
+from typing import Optional
+
+from src import variables
+from src.modules.performance_collector.codec import ProposalDuty, SyncDuty, EpochDataCodec, AttDutyMisses
+from src.types import EpochNumber
+from src.utils.range import sequence
+
+
+class DutiesDB:
+    def __init__(self, path: str):
+        self._path = path
+        self._conn = sqlite3.connect(
+            self._path, check_same_thread=False, timeout=variables.PERFORMANCE_COLLECTOR_DB_CONNECTION_TIMEOUT
+        )
+        # Optimize SQLite for performance: WAL mode for concurrent access,
+        # normal sync for speed/safety balance, memory temp storage
+        self._conn.execute("PRAGMA journal_mode=WAL;")
+        self._conn.execute("PRAGMA synchronous=NORMAL;")
+        self._conn.execute("PRAGMA temp_store=MEMORY;")
+        self._conn.execute(
+            """
+            CREATE TABLE IF NOT EXISTS duties
+            (
+                epoch INTEGER PRIMARY KEY,
+                blob  BLOB NOT NULL
+            );
+            """
+        )
+        self._conn.execute(
+            """
+            CREATE TABLE IF NOT EXISTS epochs_demand
+            (
+                consumer STRING PRIMARY KEY,
+                l_epoch INTEGER,
+                r_epoch INTEGER
+            )
+            """
+        )
+        self._conn.execute(
+            """
+            CREATE TABLE IF NOT EXISTS epochs_demand_nonce
+            (
+                value INTEGER NOT NULL
+            )
+            """
+        )
+        self._conn.execute("INSERT INTO epochs_demand_nonce (value) VALUES (0);")
+        self._conn.commit()
+
+    def __del__(self):
+        if self._conn:
+            self._conn.close()
+            self._conn = None
```

Comment:
> It is not guaranteed that `__del__()` methods are called for objects that still exist when the interpreter exits.

I'm not sure if it's a problem at all, I guess the OS will remove any file handles and free memory anyway.